app/api/logging/audit.py

```diff
@@ -20,54 +20,117 @@
-    if event_name == "open" and isinstance(args[0], str) and "/__pycache__/" in args[0]:
-        # Python has a high rate of these events in normal operation.
-        return
-    if event_name == "os.chmod" and type(args[0]) is int:
-        # Gunicorn generates a high volume of these events in normal operation (see workertmp.py)
-        return
-    if event_name == "open" and isinstance(args[0], str) and "/pytz/" in args[0]:
-        # The pytz module generates a high volume of these events in normal operation.
+def init() -> None:
+    """Initialize the audit logging module to start
+    logging security audit events."""
+    sys.addaudithook(handle_audit_event)
+
+
+def handle_audit_event(event_name: str, args: tuple[Any, ...]) -> None:
+    # Define events to log and the arguments to log for each event.
+    # For more information about these events and what they mean, see https://peps.python.org/pep-0578/#suggested-audit-hook-locations
+    # For the full list of auditable events, see https://docs.python.org/3/library/audit_events.html
+    # Define this variable locally so it can't be modified by other modules.
+    #
+    # Events from the suggested audit hook locations in PEP 578 that we aren't logging:
+    #
+    # compile and import
+    #     Detects when code is being compiled or imported.
+    #
+    #     Why we aren't logging:
+    #     Logging as part of regular imports is too noisy.
+    #
+    # sys._getframe
+    #     Detects when code is accessing frames directly.
+    #
+    #     Why we aren't logging:
+    #     The logging module itself calls getFrame
+    #     (see https://github.com/python/cpython/blob/bd7903967cd2a19ebc842dd1cce75f60a18aef02/Lib/logging/__init__.py#L170)
+    #     so logging on this event would trigger an infinite loop unless
+    #     logging._srcfile is set to None
+    #     See https://github.com/python/cpython/blob/bd7903967cd2a19ebc842dd1cce75f60a18aef02/Lib/logging/__init__.py#L179-L189
+    #
+    # object.__getattr__
+    #     Detect access to restricted attributes. This event is raised for any built-in members
+    #     that are marked as restricted, and members that may allow bypassing imports.
+    #
+    #     Why we aren't logging:
+    #     The logging module itself accesses frame.f_code.co_filename
+    #     (see https://github.com/python/cpython/blob/bd7903967cd2a19ebc842dd1cce75f60a18aef02/Lib/logging/__init__.py#L202)
+    #     which calls object.__getattr__ under the hood
+    #     (see https://github.com/python/cpython/blob/main/Objects/frameobject.c#L89-L95)
+    #
+
+    EVENTS_TO_LOG = {
+        # Detect when new audit hooks are being added.
+        "sys.addaudithook": (),
+        # Detects any attempt to set the open_code hook.
+        "cpython.PyFile_SetOpenCodeHook": (),
+        # Detect dynamic execution of code objects. This only occurs for explicit
+        # calls, and is not raised for normal function invocation.
+        "exec": ("code_object",),
```

Comment:
Was on the exclude list due to too much noise.

Reply:
any idea what causes it to be noisy? the description of it seems like something potentially worth checking for if we want to detect malicious code:

> Detect dynamic execution of code objects. This only occurs for explicit calls, and is not raised for normal function invocation.